share/wake/lib/core/json.wake

```diff
@@ -154,18 +154,155 @@ export def x // y =
       foldr flatten tail l
   JArray (helper Nil x)
 
+# Compare two JSON values for equality, as a human would judge them.
+# This includes comparing `JObject`s without regard to the internal order of
+# their keys.
 export def x ==/ y = match x y
+    (JObject a) (JObject b) =
+        def sortJObjectKeys dict =
+            def cmpKeys (Pair k1 _) (Pair k2 _) =
+                #TODO: Ensure a predictable order even when keys are equal.
+                k1 <~ k2
+            sortBy cmpKeys dict
+        def normalizeOrder =
+            JObject _.sortJObjectKeys
+        structurallyEqual (normalizeOrder a) (normalizeOrder b)
+    _ _ =
+        structurallyEqual x y
+
+# Compare two JSON values for equality according to their digital representation.
+# Note that this most significantly compares `JObject`s according to the order
+# in which their keys are stored.
+export def structurallyEqual (x: JValue) (y: JValue): Boolean = match x y
   (JString  a) (JString  b) = a ==~ b
   (JInteger a) (JInteger b) = a ==  b
   (JDouble  a) (JDouble  b) = a ==. b
   (JBoolean a) (JBoolean b) = if a then b else !b
   JNull        JNull        = True
   (JObject  a) (JObject  b) =
     def helper (Pair (Pair k c) (Pair l d)) = k ==~ l && c ==/ d
-    if a.len != b.len then False else zip a b | forall helper
+    if a.len != b.len then
+      False
+    else
+      zip a b
+      | forall helper
   (JArray   a) (JArray   b) =
     def helper (Pair c d) = c ==/ d
     if a.len != b.len then False else zip a b | forall helper
   (JArray a) _ = exists (_ ==/ y) a
   _ (JArray b) = exists (_ ==/ x) b
   _ _ = False
+
+# Simplify a JSON value according to the rules given.
+# For example, Wake's implementation of JSON is slightly more lenient than the
+# base standards, so if maximum compatibility is desired in situations where
+# some other implementation may be parsing generated output,
+# `normalizeJSONBasic` may be used to target that least common denominator.
+export def normalizeJSON (fmt: JSONNormalize): JValue => Result JValue Error = match _
+    JString a =
+        getJSONNormalizeString fmt a
+        | rmap JString
+    JInteger a =
+        getJSONNormalizeInteger fmt a
+        | rmap JInteger
+    JDouble a =
+        getJSONNormalizeDouble fmt a
+        | rmap JDouble
+    JBoolean a =
+        getJSONNormalizeBoolean fmt a
+        | rmap JBoolean
+    JNull =
+        Pass JNull
+    JObject a =
```

Comment:
Both this and the JArray cases fail to run normalizeJSON recursively after they apply the local fix-up.